Fix issue #8 by detecting message type and calling proper methods + split header validation parsing

crashdemons opened the pull request:
In OpenPGP.JS the method* to load an ascii-armor "clearsigned" message before verification is `openpgp.cleartext.readArmored` however for an ascii-armored "signed" message (not clearsign), we need to instead use `openpgp.message.readArmored`.

Additionally, the non-clearsigned version does not have a message section, just a compressed data section, similar to the signature section of a clearsigned message.  

This means three changes were needed:
 - addition of a method (PgpParsing.guessType) to determine if the message is signed or clearsigned automatically. (OpenPGP.js does not do type detection).
 - creating of a state line-parser for non-clearsigned messages (this is basically identical to the Clearsigned one, minus two lines).
 - using the type to change which OpenPGP.js method is called.

I've also included these changes:
 - Restructured files into asset directories so it's not just a big mess all in one folder.

This is intended to fix Issue #8 and causes Test-Case 24 to pass.

crashdemons committed:
reorganize project structure
(assets directory, js/css split up, etc)

crashdemons committed:
detect non-clearsigned message and use the proper OpenPGP.JS method + use different header validation for it

crashdemons committed:
change method name

crashdemons committed:
fix call to message name

tweqx commented on .gitignore:
```diff
@@ -1,3 +1,4 @@
-*.ttf
-*.otf
\ No newline at end of file
+/assets/fonts/*.ttf
```
SInce you're also adding a .gitignore file inside the /assets/font/ folder, I don't think these are necessary. But that's probably fine.

crashdemons committed:
remove unnecessary font references in .gitignore

tweqx commented on assets/js/app/pgputil.js:
```diff
@@ -326,10 +341,10 @@ var pgputil = {
 
     //checks a clearsigned armor message for discrepancies in formatting that are ignored by OpenPGP.js
     //refer to this infographic for how this function validates messages https://i.imgur.com/AvultlA.png
```
Oudated

tweqx commented on assets/js/app/pgpparsing.js:
```diff
@@ -0,0 +1,55 @@
+    },
+    guessMessageType:function(cleartext){
+        //var i1 = cleartext.indexOf("-----BEGIN PGP SIGNED MESSAGE-----");
+        var i2 = cleartext.indexOf("-----BEGIN PGP MESSAGE-----");
```
With this change, the clear-signed [MIDI PGP-message](https://pastebin.com/VuyPwuL4) now has an "invalid format", because it contains a PGP signed message (without a clear signature).

To fix this, I think the only solution is to search for `-----BEGIN PGP MESSAGE-----` (or `-----BEGIN PGP SIGNED MESSAGE-----`) at the start of lines only. You couldn't simply search for `-----BEGIN PGP SIGNED MESSAGE-----` (line 39) because even if the signature couldn't possibly contain it, I think theoretically a comment could.

crashdemons replied:
That pastebin shows "good signature" for me with the changes. the pastebin contains a space in the inner message headers. 

However you're right that this will produce incorrect results if `-----BEGIN PGP MESSAGE-----` is included in the middle of a string.  guess I will have to use lineparser, splitting, or regex after all.

tweqx replied:
>That pastebin shows "good signature" for me with the changes

Really!? ![The current changes](https://user-images.githubusercontent.com/29614762/160258527-e61ae110-544f-4936-b8e7-c71356fc537d.png)

What about doing instead:
```js
let is_signed = cleartext.startsWith("-----BEGIN PGP MESSAGE-----\n") || cleartext.indexOf("\n-----BEGIN PGP MESSAGE-----\n") !== -1;

return is_signed ? PgpParsing.parsers.SIGNED : PgpParsing.parsers.CLEARSIGNED;
```

crashdemons replied:
weird - maybe I have some local changes
![image](https://user-images.githubusercontent.com/3233238/160258923-292537c8-f5fd-45ce-889f-2955487d3d48.png)

well, issue there is that there may be content before the marker lines kick in, and I think I only warn for that.

tweqx replied:
According to the [PGP standard](https://www.rfc-editor.org/rfc/rfc4880#section-7), a valid message can't contain lines starting with dashes. These must be instead escaped by inserting a `- ` at the start of the line (= "dash escaped").
This means that a valid clear signed message will always contain exactly three lines starting with `-----` (`-----BEGIN PGP SIGNED MESSAGE-----`, `-----BEGIN PGP SIGNATURE-----`, `-----END PGP SIGNATURE-----`) and a valid (not clear) signed message will always contain with exactly of these two lines (`-----BEGIN PGP MESSAGE-----`, `-----END PGP MESSAGE-----`).

Thus, it is possible to discriminate between the message type by looking for a `-----BEGIN PGP SIGNATURE-----` line. That's exactly what I do in the code snippet I posted.

Once the message type is determined, the parsers are robust enough to be able to parse as usual the message

tweqx reviewed (dismissed):
Excepted for the mistake in the message type detection function, the pull request looks good.

Note that I didn't read the rest of the codebase (for instance, the `LineParser` class) to ensure they are correct.

crashdemons committed:
use LineParser to match the first marker line in the text, instead of indexOf
adds a test-case for the embedded message (27)
while I think this does message type detection correctly now, it does not solve the 'invalid format' error for me (which now I can reproduce...) and case 27 is failing for me

crashdemons committed:
adjust comments mentioning clearsigned messages to be more accurate

tweqx commented:
For me, all tests (but tests 25 and 26 of course) pass.
I agree with the changes regarding the "InitialLineMatcher", I still think creating a parser class just for that is overkill but as least it's cleaner this way.

crashdemons committed:
bump URI versions to prevent old caching on main site
